util/saveartifactlocally/saveartifactlocally.go

```diff
@@ -0,0 +1,123 @@
+	} else if !artifact.Target.IsRemote() && len(fromGlobMatches) <= 0 {
+		if ifExists {
+			return nil
+		}
+		return errors.Errorf("cannot save artifact %s, since it does not exist", artifact.StringCanonical())
+	}
+	isWildcard := strings.ContainsAny(fromPattern, `*?[`)
+	for _, from := range fromGlobMatches {
+		fiSrc, err := os.Stat(from)
+		if err != nil {
+			return errors.Wrapf(err, "os stat %s", from)
+		}
+		srcIsDir := fiSrc.IsDir()
+		to := destPath
+		destIsDir := strings.HasSuffix(to, "/") || to == "."
+		if artifact.Target.IsLocalExternal() && !filepath.IsAbs(to) {
+			// Place within external dir.
+			to = path.Join(artifact.Target.LocalPath, to)
+		}
+		if destIsDir {
+			// Place within dest dir.
+			to = path.Join(to, path.Base(from))
+		}
+		destExists := false
+		fiDest, err := os.Stat(to)
+		if err != nil {
+			// Ignore err. Likely dest path does not exist.
+			if isWildcard && !destIsDir {
+				return errors.New(
+					"artifact is a wildcard, but AS LOCAL destination does not end with /")
+			}
+			destIsDir = fiSrc.IsDir()
+		} else {
+			destExists = true
+			destIsDir = fiDest.IsDir()
+		}
+		if srcIsDir && !destIsDir {
+			return errors.New(
+				"artifact is a directory, but existing AS LOCAL destination is a file")
+		}
+		if destExists {
+			if !srcIsDir {
+				// Remove preexisting dest file.
+				err = os.Remove(to)
+				if err != nil {
+					return errors.Wrapf(err, "rm %s", to)
+				}
+			} else {
+				// Remove preexisting dest dir.
+				err = os.RemoveAll(to)
+				if err != nil {
+					return errors.Wrapf(err, "rm -rf %s", to)
+				}
+			}
+		}
+
+		toDir := path.Dir(to)
+		err = os.MkdirAll(toDir, 0755)
+		if err != nil {
+			return errors.Wrapf(err, "mkdir all for artifact %s", toDir)
+		}
+		err = os.Link(from, to)
+		if err != nil {
+			// Hard linking did not work. Try recursive copy.
+			errCopy := reccopy.Copy(from, to)
+			if errCopy != nil {
+				return errors.Wrapf(errCopy, "copy artifact %s", from)
+			}
+		}
+
+		// Write to console about this artifact.
+		artifactPath := trimFilePathPrefix(indexOutDir, from, parentConsole)
+		artifact2 := domain.Artifact{
+			Target:   artifact.Target,
+			Artifact: artifactPath,
+		}
+		destPath2 := filepath.FromSlash(destPath)
+		if strings.HasSuffix(destPath, "/") {
+			destPath2 = filepath.Join(destPath2, filepath.Base(artifactPath))
+		}
+		if printPhases {
+			artifactColor := parentConsole.WithPrefixAndSalt(artifact.Target.String(), salt).PrefixColor()
+			artifactStr := artifactColor.Sprintf("%s", artifact2.StringCanonical())
+			console.Printf("Artifact %s output as %s\n", artifactStr, destPath2)
+		}
+	}
+	return nil
+}
+
+func trimFilePathPrefix(prefix string, thePath string, console conslogging.ConsoleLogger) string {
+	ret, err := filepath.Rel(prefix, thePath)
+	if err != nil {
+		console.Warnf("Warning: Could not compute relative path for %s "+
+			"as being relative to %s: %s\n", thePath, prefix, err.Error())
+		return thePath
+	}
+	return ret
+}
```

Comment:
This code was simply moved without modification, right?

Reply:
yes (with exception for `parentConsole` now being passed in):

```
$ diff old new
1c1,2
< func (b *Builder) saveArtifactLocally(ctx context.Context, console *conslogging.BufferedLogger, artifact domain.Artifact, indexOutDir string, destPath string, salt string, opt BuildOpt, ifExists bool) error {
---
> // SaveArtifactLocally handles saving artifacts to the local host, and is called from both builder and waitblock
> func SaveArtifactLocally(ctx context.Context, parentConsole conslogging.ConsoleLogger, console *conslogging.BufferedLogger, artifact domain.Artifact, indexOutDir string, destPath string, salt string, printPhases, ifExists bool) error {
80c81
< 		artifactPath := trimFilePathPrefix(indexOutDir, from, b.opt.Console)
---
> 		artifactPath := trimFilePathPrefix(indexOutDir, from, parentConsole)
89,90c90,91
< 		if opt.PrintPhases {
< 			artifactColor := b.opt.Console.WithPrefixAndSalt(artifact.Target.String(), salt).PrefixColor()
---
> 		if printPhases {
> 			artifactColor := parentConsole.WithPrefixAndSalt(artifact.Target.String(), salt).PrefixColor()
```